sql/ttl/ttljob: TestRowLevelTTLJobRandomEntries failed

sql/ttl/ttljob.TestRowLevelTTLJobRandomEntries [failed](https://mesolite.cluster.engflow.com/invocations/default/2cd5cc21-abd0-43b7-a3fa-ae1b2d26d7c9?testReportRun=1&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvc3FsL3R0bC90dGxqb2I6dHRsam9iX3Rlc3Q=) on master @ [f8a09a5c1dbcbfcb537ba4050c6061ef6d917687](https://github.com/cockroachdb/cockroach/commits/f8a09a5c1dbcbfcb537ba4050c6061ef6d917687):


```
=== RUN   TestRowLevelTTLJobRandomEntries
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestRowLevelTTLJobRandomEntries2244555307
    test_log_scope.go:82: use -show-logs to present logs inline
    ttljob_test.go:1043: -- test log scope end --
test logs left over in: outputs.zip/logTestRowLevelTTLJobRandomEntries2244555307
--- FAIL: TestRowLevelTTLJobRandomEntries (191.32s)
=== RUN   TestRowLevelTTLJobRandomEntries/random_1
    ttljob_test.go:947: test case (runAtMixedClusterVersion=true): ttljob_test.testCase{desc:"random 1", createTable:"CREATE TABLE tbl (\n\tid UUID DEFAULT gen_random_uuid(),\n\trand_col_1 LTREE,\n\trand_col_2 TIMESTAMP,\n\tt TEXT NULL,\n\ti INT8 NULL,\n\t\n\tPRIMARY KEY (id, rand_col_1, rand_col_2)\n) WITH (ttl_expire_after = '30 days', ttl_select_batch_size = 16, ttl_delete_batch_size = 22)", preSetup:[]string(nil), postSetup:[]string(nil), numExpiredRows:1115, numNonExpiredRows:27, numSplits:4, forceNonMultiTenant:false, expirationExpression:"", addRow:(func(*ttljob_test.rowLevelTTLTestJobTestHelper, *testing.T, *tree.CreateTable, time.Time))(nil)}
    test_server_shim.go:92: cluster virtualization disabled due to issue: #109391 (expected label: C-bug)
    ttljob_test.go:964: error executing query="CREATE TABLE tbl (\n\tid UUID DEFAULT gen_random_uuid(),\n\trand_col_1 LTREE,\n\trand_col_2 TIMESTAMP,\n\tt TEXT NULL,\n\ti INT8 NULL,\n\t\n\tPRIMARY KEY (id, rand_col_1, rand_col_2)\n) WITH (ttl_expire_after = '30 days', ttl_select_batch_size = 16, ttl_delete_batch_size = 22)" args=[]: pq: ltree not supported until version 25.4
--- FAIL: TestRowLevelTTLJobRandomEntries/random_1 (0.34s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRowLevelTTLJobRandomEntries.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-53795

sql/ttl/ttljob.TestRowLevelTTLJobRandomEntries [failed](https://mesolite.cluster.engflow.com/invocations/default/7ab32420-06ec-4aa9-915d-6565ac079a94?testReportRun=3&testReportShard=2&testReportAttempt=1#targets-Ly9wa2cvc3FsL3R0bC90dGxqb2I6dHRsam9iX3Rlc3Q=) on master @ [f8a09a5c1dbcbfcb537ba4050c6061ef6d917687](https://github.com/cockroachdb/cockroach/commits/f8a09a5c1dbcbfcb537ba4050c6061ef6d917687):


```
=== RUN   TestRowLevelTTLJobRandomEntries
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestRowLevelTTLJobRandomEntries8105950
    test_log_scope.go:82: use -show-logs to present logs inline
    ttljob_test.go:1043: -- test log scope end --
test logs left over in: outputs.zip/logTestRowLevelTTLJobRandomEntries8105950
--- FAIL: TestRowLevelTTLJobRandomEntries (196.10s)
=== RUN   TestRowLevelTTLJobRandomEntries/random_4
    ttljob_test.go:947: test case (runAtMixedClusterVersion=true): ttljob_test.testCase{desc:"random 4", createTable:"CREATE TABLE tbl (\n\tid UUID DEFAULT gen_random_uuid(),\n\trand_col_1 PG_LSN,\n\trand_col_2 LTREE,\n\tt TEXT NULL,\n\ti INT8 NULL,\n\tFAMILY fam0 (t, i, rand_col_1, rand_col_2, id), \n\tPRIMARY KEY (id, rand_col_1, rand_col_2)\n) WITH (ttl_expire_after = '30 days', ttl_select_batch_size = 66, ttl_delete_batch_size = 88)", preSetup:[]string(nil), postSetup:[]string(nil), numExpiredRows:287, numNonExpiredRows:17, numSplits:8, forceNonMultiTenant:false, expirationExpression:"", addRow:(func(*ttljob_test.rowLevelTTLTestJobTestHelper, *testing.T, *tree.CreateTable, time.Time))(nil)}
    test_server_shim.go:92: cluster virtualization disabled due to issue: #109391 (expected label: C-bug)
    ttljob_test.go:964: error executing query="CREATE TABLE tbl (\n\tid UUID DEFAULT gen_random_uuid(),\n\trand_col_1 PG_LSN,\n\trand_col_2 LTREE,\n\tt TEXT NULL,\n\ti INT8 NULL,\n\tFAMILY fam0 (t, i, rand_col_1, rand_col_2, id), \n\tPRIMARY KEY (id, rand_col_1, rand_col_2)\n) WITH (ttl_expire_after = '30 days', ttl_select_batch_size = 66, ttl_delete_batch_size = 88)" args=[]: pq: ltree not supported until version 25.4
--- FAIL: TestRowLevelTTLJobRandomEntries/random_4 (0.35s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=3</code>
 - <code>shard=2</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRowLevelTTLJobRandomEntries.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql/ttl/ttljob.TestRowLevelTTLJobRandomEntries [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20383539?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20383539?buildTab=artifacts#/) on master @ [c74f6a36b0ab9ee6471171400249258d2ba7b292](https://github.com/cockroachdb/cockroach/commits/c74f6a36b0ab9ee6471171400249258d2ba7b292):


```
=== RUN   TestRowLevelTTLJobRandomEntries
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/2369c83d90b731358db5a0b39ab9f352/logTestRowLevelTTLJobRandomEntries3318880453
    test_log_scope.go:82: use -show-logs to present logs inline
    ttljob_test.go:1043: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/2369c83d90b731358db5a0b39ab9f352/logTestRowLevelTTLJobRandomEntries3318880453
--- FAIL: TestRowLevelTTLJobRandomEntries (186.72s)
=== RUN   TestRowLevelTTLJobRandomEntries/random_4
    ttljob_test.go:947: test case (runAtMixedClusterVersion=true): ttljob_test.testCase{desc:"random 4", createTable:"CREATE TABLE tbl (\n\tid UUID DEFAULT gen_random_uuid(),\n\trand_col_1 LTREE,\n\trand_col_2 BOOL,\n\tt TEXT NULL,\n\ti INT8 NULL,\n\tFAMILY fam0 (rand_col_1, i), \n\tPRIMARY KEY (id, rand_col_1, rand_col_2)\n) WITH (ttl_expire_after = '30 days', ttl_select_batch_size = 77, ttl_delete_batch_size = 92)", preSetup:[]string(nil), postSetup:[]string(nil), numExpiredRows:703, numNonExpiredRows:33, numSplits:8, forceNonMultiTenant:false, expirationExpression:"", addRow:(func(*ttljob_test.rowLevelTTLTestJobTestHelper, *testing.T, *tree.CreateTable, time.Time))(nil)}
    test_server_shim.go:92: cluster virtualization disabled due to issue: #109391 (expected label: C-bug)
    ttljob_test.go:964: error executing query="CREATE TABLE tbl (\n\tid UUID DEFAULT gen_random_uuid(),\n\trand_col_1 LTREE,\n\trand_col_2 BOOL,\n\tt TEXT NULL,\n\ti INT8 NULL,\n\tFAMILY fam0 (rand_col_1, i), \n\tPRIMARY KEY (id, rand_col_1, rand_col_2)\n) WITH (ttl_expire_after = '30 days', ttl_select_batch_size = 77, ttl_delete_batch_size = 92)" args=[]: pq: ltree not supported until version 25.4
--- FAIL: TestRowLevelTTLJobRandomEntries/random_4 (0.55s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRowLevelTTLJobRandomEntries.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql/ttl/ttljob.TestRowLevelTTLJobRandomEntries [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20387441?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/20387441?buildTab=artifacts#/) on master @ [e46fe1bb99999ff30e63cf306c5324737b10b79a](https://github.com/cockroachdb/cockroach/commits/e46fe1bb99999ff30e63cf306c5324737b10b79a):


```
=== RUN   TestRowLevelTTLJobRandomEntries
    test_log_scope.go:171: test logs captured to: /artifacts/tmp/_tmp/2369c83d90b731358db5a0b39ab9f352/logTestRowLevelTTLJobRandomEntries2981479403
    test_log_scope.go:82: use -show-logs to present logs inline
    ttljob_test.go:1043: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/2369c83d90b731358db5a0b39ab9f352/logTestRowLevelTTLJobRandomEntries2981479403
--- FAIL: TestRowLevelTTLJobRandomEntries (187.83s)
=== RUN   TestRowLevelTTLJobRandomEntries/random_5
    ttljob_test.go:947: test case (runAtMixedClusterVersion=true): ttljob_test.testCase{desc:"random 5", createTable:"CREATE TABLE tbl (\n\tid UUID DEFAULT gen_random_uuid(),\n\trand_col_1 STRING,\n\trand_col_2 LTREE,\n\tt TEXT NULL,\n\ti INT8 NULL,\n\tFAMILY fam0 (t, id, rand_col_1, i), FAMILY fam1 (rand_col_2), \n\tPRIMARY KEY (id, rand_col_1, rand_col_2)\n) WITH (ttl_expire_after = '30 days', ttl_select_batch_size = 82, ttl_delete_batch_size = 92)", preSetup:[]string(nil), postSetup:[]string(nil), numExpiredRows:1025, numNonExpiredRows:5, numSplits:9, forceNonMultiTenant:false, expirationExpression:"", addRow:(func(*ttljob_test.rowLevelTTLTestJobTestHelper, *testing.T, *tree.CreateTable, time.Time))(nil)}
    test_server_shim.go:92: cluster virtualization disabled due to issue: #109391 (expected label: C-bug)
    ttljob_test.go:964: error executing query="CREATE TABLE tbl (\n\tid UUID DEFAULT gen_random_uuid(),\n\trand_col_1 STRING,\n\trand_col_2 LTREE,\n\tt TEXT NULL,\n\ti INT8 NULL,\n\tFAMILY fam0 (t, id, rand_col_1, i), FAMILY fam1 (rand_col_2), \n\tPRIMARY KEY (id, rand_col_1, rand_col_2)\n) WITH (ttl_expire_after = '30 days', ttl_select_batch_size = 82, ttl_delete_batch_size = 92)" args=[]: pq: ltree not supported until version 25.4
--- FAIL: TestRowLevelTTLJobRandomEntries/random_5 (0.42s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRowLevelTTLJobRandomEntries.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/ttl/ttljob/ttljob_test.go
+++ b/pkg/sql/ttl/ttljob/ttljob_test.go
@@ -9,7 +9,6 @@ import (
 	"bytes"
 	"context"
 	"fmt"
-	"math/rand"
 	"regexp"
 	"strconv"
 	"strings"
@@ -702,9 +701,17 @@ func TestRowLevelTTLJobRandomEntries(t *testing.T) {
 		// causes parsing errors.
 		case types.CollatedStringFamily:
 			if typ.Oid() != oidext.T_citext && typ.Oid() != oidext.T__citext {
-				// TODO(foundations): CITEXT should only be prohibited when
-				// running with mixed 25.2 version where the type is not
-				// supported.
+				if int(clusterversion.MinSupported) >= int(clusterversion.V25_3) {
+					// CITEXT is only supported in 25.3+, so if we happen to run
+					// the test in the mixed version variant, we can't use the
+					// type.
+					indexableTyps = append(indexableTyps, typ)
+				}
+			}
+		case types.LTreeFamily:
+			if int(clusterversion.MinSupported) >= int(clusterversion.V25_4) {
+				// LTREE is only supported in 25.4+, so if we happen to run the
+				// test in the mixed version variant, we can't use the type.
 				indexableTyps = append(indexableTyps, typ)
 			}
 		default:
@@ -956,7 +963,7 @@ func TestRowLevelTTLJobRandomEntries(t *testing.T) {
 
 	for _, tc := range testCases {
 		t.Run(tc.desc, func(t *testing.T) {
-			runAtMixedClusterVersion := rand.Intn(2) == 0
+			runAtMixedClusterVersion := rng.Intn(2) == 0
 			// Log to make it slightly easier to reproduce a random config.
 			t.Logf("test case (runAtMixedClusterVersion=%t): %#v", runAtMixedClusterVersion, tc)
 			th, cleanupFunc := newRowLevelTTLTestJobTestHelper(

